sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java

```diff
@@ -62,310 +58,272 @@
-      checkArgument(maxAttempts > 0, "maxAttempts should be greater than 0");
-      checkArgument(
-          maxDuration != null && maxDuration.isLongerThan(Duration.ZERO),
-          "maxDuration should be greater than 0");
-      return new AutoValue_SnsIO_RetryConfiguration.Builder()
-          .setMaxAttempts(maxAttempts)
-          .setMaxDuration(maxDuration)
-          .setRetryPredicate(DEFAULT_RETRY_PREDICATE)
-          .build();
-    }
-
-    @AutoValue.Builder
-    abstract static class Builder {
-      abstract Builder setMaxAttempts(int maxAttempts);
-
-      abstract Builder setMaxDuration(Duration maxDuration);
-
-      abstract Builder setRetryPredicate(RetryPredicate retryPredicate);
-
-      abstract RetryConfiguration build();
-    }
-
-    /**
-     * An interface used to control if we retry the SNS Publish call when a {@link Throwable}
-     * occurs. If {@link RetryPredicate#test(Object)} returns true, {@link Write} tries to resend
-     * the requests to the Solr server if the {@link RetryConfiguration} permits it.
-     */
-    @FunctionalInterface
-    interface RetryPredicate extends Predicate<Throwable>, Serializable {}
-
-    private static class DefaultRetryPredicate implements RetryPredicate {
-      private static final ImmutableSet<Integer> ELIGIBLE_CODES =
-          ImmutableSet.of(HttpStatus.SC_SERVICE_UNAVAILABLE);
-
-      @Override
-      public boolean test(Throwable throwable) {
-        return (throwable instanceof IOException
-            || (throwable instanceof InternalErrorException)
-            || (throwable instanceof InternalErrorException
-                && ELIGIBLE_CODES.contains(((InternalErrorException) throwable).statusCode())));
-      }
-    }
-  }
-
-  /** Implementation of {@link #write}. */
+  /** Implementation of {@link #write()}. */
   @AutoValue
   public abstract static class Write<T>
-      extends PTransform<PCollection<T>, PCollection<PublishResponse>> {
-    @Nullable
-    abstract String getTopicArn();
+      extends PTransform<PCollection<T>, PCollection<WrappedSnsResponse<T>>> {
 
     @Nullable
-    abstract SerializableFunction<T, PublishRequest> getPublishRequestFn();
+    abstract SnsClientProvider getSnsClientProvider();
+
+    /** Flag to indicate if io should throw an error on exception. */
+    abstract boolean getThrowOnError();
 
+    /** Coder for element T. */
     @Nullable
-    abstract SnsClientProvider getSnsClientProvider();
+    abstract Coder<T> getElementCoder();
 
+    /** SerializableFunction to create PublishRequest. */
     @Nullable
-    abstract RetryConfiguration getRetryConfiguration();
+    abstract SerializableFunction<T, PublishRequest> getPublishRequestFn();
 
     abstract Builder<T> builder();
 
     @AutoValue.Builder
     abstract static class Builder<T> {
+      abstract Builder<T> setSnsClientProvider(SnsClientProvider clientProvider);
 
-      abstract Builder<T> setTopicArn(String topicArn);
```

Comment:
How do we specify topic(s) if `getTopicArn/setTopicArn` are not used anymore?

Reply:
Hey, it can be done using `setPublishRequestFn`, something like this-
```message -> PublishRequest.builder().message(message).topicArn(topicArn).build()```

Reply:
What if `publishRequestFn` won't be specified? If it's required then we need to provide either  default implementation or warn user about that and don't run a pipeline.